command/server_dev.go

```diff
@@ -0,0 +1,900 @@
+		policies []string
+	}
+
+	authAppRole struct {
+		roleName string
+		roleID   string
+		secretID string
+		policies []string
+	}
+
+	authUserpass struct {
+		username string
+		password string
+		policies []string
+	}
+
+	secrets []quickstartSecret
+
+	policyMap map[string]string
+}
+
+type quickstartSecret struct {
+	key            string
+	data           map[string]interface{}
+	customMetadata map[string]interface{}
+}
+
+func enableQuickstart(ctx context.Context, init *vault.InitResult, core *vault.Core) (*quickstartOptions, error) {
+	readOnlyPolicy := "path \"secret/*\" {capabilities = [\"read\"]}"
+	readOnlyPolicyName := "read-only"
+	crudPolicy := "path \"secret/*\" {capabilities = [\"create\", \"read\", \"update\", \"delete\"]}"
+	crudPolicyName := "developer"
+
+	policies := make(map[string]string)
+	policies[crudPolicyName] = crudPolicy
+	policies[readOnlyPolicyName] = readOnlyPolicy
+
+	options := &quickstartOptions{
+		policyMap: policies,
+
+		identity: struct {
+			name     string
+			entityID string
+			policies []string
+		}{name: "demo-identity", policies: []string{crudPolicyName}},
+
+		authAppRole: struct {
+			roleName string
+			roleID   string
+			secretID string
+			policies []string
+		}{roleName: "app-read-only", policies: []string{readOnlyPolicyName}},
+
+		authUserpass: struct {
+			username string
+			password string
+			policies []string
+		}{username: "demo-username", password: "superSecretDemo", policies: nil},
+
+		secrets: []quickstartSecret{{
+			key: "sample-secret",
+			data: map[string]interface{}{
+				"foo": "bar",
+				"zip": "zap",
+			},
+			customMetadata: map[string]interface{}{
+				"team": "demo",
+			},
+		}},
+	}
+
+	// add policies
+	for name, policy := range options.policyMap {
+		if err := devAddPolicy(ctx, init, core, name, policy); err != nil {
+			return nil, err
+		}
+	}
+
+	// userpass
+	err := qsEnableUserpass(ctx, init, core, options)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initalize quickstart userpass: %w", err)
+	}
+
+	// app role auth
+	err = qsEnableAppRole(ctx, init, core, options)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initalize quickstart approle: %w", err)
+	}
+
+	// identity
+	err = qsEnableIdentity(ctx, init, core, options)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initalize quickstart identity: %w", err)
+	}
+
+	// sample secret
+	for _, s := range options.secrets {
+		err = devAddSecret(ctx, init, core, s)
+		if err != nil {
+			return nil, fmt.Errorf("failed to initalize quickstart seeded secret: %w", err)
+		}
+	}
+
+	return options, nil
+}
+
+func devAddSecret(ctx context.Context, init *vault.InitResult, core *vault.Core, secret quickstartSecret) error {
+	r := &logical.Request{
+		Operation:   logical.UpdateOperation,
+		ClientToken: init.RootToken,
+		Path:        fmt.Sprintf("secret/data/%s", secret.key),
+		Data: map[string]interface{}{
+			"data": secret.data,
+		},
+	}
+	resp, err := core.HandleRequest(ctx, r)
```

Comment:
Just for my own curiosity since I haven't stumbled across this before, what is the difference between using this method and Logical().Write in the API package? (since I think I've seen other code in `command/` using that, or at least Logical().RawRequest)

Reply:
This method appears to be the pattern folks use when interacting with server-side resources in situations like this, as opposed to the client-side (which I suppose a good portion of `command` utilizes) where `(*Client).Logical().Write()` is more appropriate. [Example](https://github.com/hashicorp/vault/blob/06c56355b59c14a950de627ac3724769bb3c0834/command/server.go#L1754-L1773)